src/betterproto/plugin/models.py

```diff
@@ -281,6 +284,42 @@ def __post_init__(self):
         self.deprecated = self.proto_obj.options.deprecated
         super().__post_init__()
 
+    def compile(self, indent_level: int = 0) -> str:
+        """Construct Python code from MessageCompiler instance."""
+        out = f"class {self.py_name}(betterproto.Message):"
```

Comment:
To be clear this is OK, certainly not worse than the jinja template. Building up a string line by line is clear enough, if a bit visually noisy, and a bit inefficient with all that copying. 

However I wonder if it could be clearer and more concise by building up and joining a list like:

```python
result = [f"class {self.py_name}(betterproto.Message):"]

if self.comment:
    result.append(self.comment)

if self.fields:
    result.append(
        field.compile(indent_level=indent_level + 1) for field in self.fields
    )
...

return "\n".join(result)
```

Or even get a little declarative DSL going to manage the indentation and compiling of child compilers like:

```python
code = CodeBuilder(indent=indent_level)
code(f"class {self.py_name}(betterproto.Message):")

if self.comment:
    code(self.comment, space=1) # space adds extra empty lines after

if self.fields:
    code(*self.fields, indent=1) # *args can be Compilers, and indents are cumulative
else:
    code("pass", indent=1)

if self.deprecated or list(self.deprecated_fields):
    code("def __post_init__(self) -> None:", indent=1)
    code("super().__post_init__()", indent=2)
...
return code.build()
```

Given a "simple" helper like:
```python
class CodeBuilder:
    def __init__(self, indent:str):
        self._base_indent = indent
        self.lines = []

    def __call__(*lines, *, indent: int = 0, space: int = 0)
        ...

    def compile(): str
        ...
```

I'm not saying it's necessary, and there's a risk of a getting carried away, but it could be nice. What do you think?

Reply:
I like the `CodeBuilder` because it can be used to clean up new lines and track indentations. I think that as long as it's clear that it's only function is building the code and formatting (i.e. it has no knowledge of what the lines of code it's being fed mean) the scope should be manageable.

Reply:
Regarding performance, while we should obviously try to optimize what can be optimized, my personal feeling is that speed here should not be as much of a concern as clear bug-free code.